### PR TITLE
UofT-DSI | production - Assignment 2

### PR DESCRIPTION
**Pull request by slathwal**
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
- I am writing code to to build a machine learning model to predict the area burnt by forest fires. 

## What did you learn from the changes you have made?
I learned the following in this assignment:
- Testing several ways of pre_processing your data, linear and non-linear transformations, scaling, one-hot encoding etc.
- Putting data pre-processing steps into pipelines using Pipelines and Column Transformers in scikit-learn
- Combining the data pre-processing and machine learning models into a single pipeline
- Tuning the entire pipeline (data pre-processing+ML model) together with GridSearchCV
- Using various scoring methods in GridSearchCV
- Comparing several pipelines and choosing the best one based on a metric like root mean squared error
- Saving the best model as a pickle file for use later
- Testing feature importance for individual observations as well as for the entire dataset using SHAP values

## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?
NA

## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?
Yes, some GridSearchCV pipelines took a very long time (>40 mins). It was not clear which hyper parameters to tune that would have the greatest impact on performance. It is also unclear how to choose the best scoring method for a problem. Since this was a regression problem, I used root mean squared error.

## How were these changes tested?
NA

## A reference to a related issue in your repository (if applicable)
NA

## Checklist
- [x] I can confirm that my changes are working as intended

### CODE REVIEW

**Review by xindizhang — APPROVED** [review]
**Great work - No further changes needed.** 
Amaing PR description. Amazing data exploration. Well-written and well-organized pipeline! You’ve met all the requirements outlined in the assignment instructions
Some minor suggestions for the future:
- The range (1–50) for hyperparameter options may be large and slow; consider stepping (e.g., range(1, 50, 5)) for efficiency.

**Comment by slathwal** [comment]
@xindizhang Thank you so much for your feedback. Your point on stepping is noted. I will now merge the PR.
